File: extensions/motor28byj48/src/lib/stepper_motor.py
import asyncio
import time

from machine import Pin
from micropython import const

FULL_ROTATION = (
    4075.7728395061727 / 8
)  # http://www.jangeox.be/2013/10/stepper-motor-28byj-48_25.html


class StepperMotor:
    MODE_NONE = const(0)
    MODE_RUN_FORWARD = const(1)
    MODE_RUN_BACKWARD = const(2)
    MODE_RUN_WITH_POSITION = const(3)

    SINGLE_PHASE_FULL_STEP_EXCITATION = const(0)
    TWO_PHASE_FULL_STEP_EXCITATION = const(1)
    HALF_PHASE_EXCITATION = const(2)

    STEPS = (
        (  # A, B, C, D
            (1, 0, 0, 0),
            (0, 1, 0, 0),
            (0, 0, 1, 0),
            (0, 0, 0, 1),
        ),
        (
            (1, 1, 0, 0),
            (0, 1, 1, 0),
            (0, 0, 1, 1),
            (1, 0, 0, 1),
        ),
        (
            (1, 0, 0, 0),
            (1, 1, 0, 0),
            (0, 1, 0, 0),
            (0, 1, 1, 0),
            (0, 0, 1, 0),
            (0, 0, 1, 1),
            (0, 0, 0, 1),
            (1, 0, 0, 1),
        ),
    )

    def __init__(self, pin_a, pin_b, pin_c, pin_d) -> None:
        self._pin_ins = (
            Pin(pin_a, Pin.OUT),
            Pin(pin_b, Pin.OUT),
            Pin(pin_c, Pin.OUT),
            Pin(pin_d, Pin.OUT),
        )
        for pin in self._pin_ins:
            pin.value(1)
        self._excitation = StepperMotor.HALF_PHASE_EXCITATION
        self._current_step = 0
        self._target_step = 0
        self._mode = StepperMotor.MODE_NONE
        self._last_step_time_us = -1
        self._rpm = 10
        self._step_us = 2000
        self._step_values = StepperMotor.STEPS[self._excitation]
        self._update_step_us()

    @property
    def excitation(self):
        return self._excitation

    @excitation.setter
    def excitation(self, excitation):
        self._excitation = excitation
        self._step_values = StepperMotor.STEPS[self._excitation]
        self._update_step_us()

    @property
    def rpm(self):
        return self._rpm

    @rpm.setter
    def rpm(self, rpm):
        if rpm <= 0:
            raise ValueError(f"{rpm} is not rpm value")

        self._rpm = float(rpm)
        self._update_step_us()

    def _update_step_us(self):
        self._step_us = (60 * 1000 * 1000) / (
            self._rpm * FULL_ROTATION * len(self._step_values)
        )

    def _angle_to_step(self, angle):
        return round(angle * FULL_ROTATION * len(self._step_values) / 360)

    def _step_to_angle(self, step):
        return step * 360.0 / len(self._step_values) / FULL_ROTATION

    @property
    def reached_target_angle(self):
        return self._mode == StepperMotor.MODE_NONE

    @property
    def current_step(self):
        return self._current_step

    @current_step.setter
    def current_step(self, current_step):
        self._current_step = current_step

    @property
    def current_angle(self):
        return self._step_to_angle(self._current_step)

    @current_angle.setter
    def current_angle(self, current_angle):
        self._current_step = self._angle_to_step(current_angle)

    @property
    def target_step(self):
        return self._target_step

    @target_step.setter
    def target_step(self, target_step):
        self._mode = StepperMotor.MODE_RUN_WITH_POSITION
        self._target_step = float(target_step)

    def move(self, steps):
        self.target_step = self.current_step - steps

    @property
    def target_angle(self):
        return self._target_angle

    @target_angle.setter
    def target_angle(self, target_angle):
        self._target_angle = float(target_angle)
        self._mode = StepperMotor.MODE_RUN_WITH_POSITION
        self._target_step = self._angle_to_step(self._target_angle)

    def rotate(self, angle):
        self.target_angle = self.current_angle - angle

    def forward(self):
        self._mode = StepperMotor.MODE_RUN_FORWARD

    def backward(self):
        self._mode = StepperMotor.MODE_RUN_BACKWARD

    def stop(self):
        self._mode = StepperMotor.MODE_NONE
        for pin in self._pin_ins:
            pin.value(1)

    def tick(self):
        if (
            self._last_step_time_us != -1
            and time.ticks_us() - self._last_step_time_us < self._step_us
        ):
            return

        self._last_step_time_us = time.ticks_us()

        if self._mode == StepperMotor.MODE_RUN_WITH_POSITION:
            if self._target_step > self._current_step:
                self._current_step += 1
            elif self._target_step < self._current_step:
                self._current_step -= 1
            if self._current_step == self._target_step:
                self._mode = StepperMotor.MODE_NONE
        elif self._mode == StepperMotor.MODE_RUN_FORWARD:
            self._current_step -= 1
        elif self._mode == StepperMotor.MODE_RUN_BACKWARD:
            self._current_step += 1

        current_step = self._current_step % len(self._step_values)
        if current_step < 0:
            current_step += len(self._step_values)

        step_value = self._step_values[current_step]
        for i in range(4):
            self._pin_ins[i].value(step_value[i])

    async def run_wait(self):
        while self._mode != StepperMotor.MODE_NONE:
            self.tick()
            await asyncio.sleep(0)

    def run(self):
        asyncio.create_task(self.run_wait())

File: extensions/motor28byj48/src/lib/files/stepper_motor.cpp
#include "stepper_motor.h"

// 步进序列定义
const int StepperMotor::STEP_SEQUENCE[3][8][4] = {
    {// 单相全步励磁
     {1, 0, 0, 0},
     {0, 1, 0, 0},
     {0, 0, 1, 0},
     {0, 0, 0, 1},
     {0, 0, 0, 0}, // 填充，使所有序列都有8个元素
     {0, 0, 0, 0},
     {0, 0, 0, 0},
     {0, 0, 0, 0}},
    {// 双相全步励磁
     {1, 1, 0, 0},
     {0, 1, 1, 0},
     {0, 0, 1, 1},
     {1, 0, 0, 1},
     {0, 0, 0, 0}, // 填充
     {0, 0, 0, 0},
     {0, 0, 0, 0},
     {0, 0, 0, 0}},
    {// 半步励磁
     {1, 0, 0, 0},
     {1, 1, 0, 0},
     {0, 1, 0, 0},
     {0, 1, 1, 0},
     {0, 0, 1, 0},
     {0, 0, 1, 1},
     {0, 0, 0, 1},
     {1, 0, 0, 1}}};

// 构造函数
StepperMotor::StepperMotor(int pin_a, int pin_b, int pin_c, int pin_d) {
  pinA = pin_a;
  pinB = pin_b;
  pinC = pin_c;
  pinD = pin_d;

  // 初始化引脚
  pinMode(pinA, OUTPUT);
  pinMode(pinB, OUTPUT);
  pinMode(pinC, OUTPUT);
  pinMode(pinD, OUTPUT);

  // 初始状态：所有引脚为低
  digitalWrite(pinA, LOW);
  digitalWrite(pinB, LOW);
  digitalWrite(pinC, LOW);
  digitalWrite(pinD, LOW);

  _excitation = HALF_PHASE_EXCITATION;
  _current_step = 0;
  _target_step = 0;
  _mode = MODE_NONE;
  _rpm = 10;
  _last_step_time = 0;
  _target_angle = 0.0;
  _current_sequence_step = 0;

  updateStepDelay();
}

// 析构函数
StepperMotor::~StepperMotor() {
  // 停止电机
  stop();
}

// 设置引脚状态
void StepperMotor::setPins(int a, int b, int c, int d) {
  digitalWrite(pinA, a);
  digitalWrite(pinB, b);
  digitalWrite(pinC, c);
  digitalWrite(pinD, d);
}

// 获取励磁模式
int StepperMotor::getExcitation() { return _excitation; }

// 设置励磁模式
void StepperMotor::setExcitation(int excitation) {
  if (excitation >= 0 && excitation <= 2) {
    _excitation = excitation;
    _current_sequence_step = 0;
    updateStepDelay();
  }
}

// 获取转速
int StepperMotor::getRpm() { return _rpm; }

// 设置转速
void StepperMotor::setRpm(int rpm) {
  if (rpm > 0) {
    _rpm = rpm;
    updateStepDelay();
  }
}

// 更新步进延迟时间
void StepperMotor::updateStepDelay() {
  if (_rpm > 0) {
    // 计算每步的延迟时间（毫秒）
    // 对于28BYJ-48电机，使用半步模式（8步序列）
    // 完整一周需要4096步（64 * 64）
    // 计算：每分钟转数 * 每转步数 = 每分钟总步数
    // 转换为毫秒/步：60000 / (rpm * steps_per_rev)

    // 根据励磁模式确定每转步数
    int steps_per_sequence;
    switch (_excitation) {
    case SINGLE_PHASE_FULL_STEP_EXCITATION:
      steps_per_sequence = 4;
      break;
    case TWO_PHASE_FULL_STEP_EXCITATION:
      steps_per_sequence = 4;
      break;
    case HALF_PHASE_EXCITATION:
    default:
      steps_per_sequence = 8;
      break;
    }

    // 实际每转步数 = 4096 / (8 / steps_per_sequence)
    // 这是因为4096是假设使用8步序列时的步数
    float actual_steps_per_rev =
        FULL_ROTATION_STEPS * (steps_per_sequence / 8.0);

    float steps_per_minute = _rpm * actual_steps_per_rev;
    _step_delay_ms = round(60000.0 / steps_per_minute);

    // 确保最小延迟
    if (_step_delay_ms < 1) {
      _step_delay_ms = 1;
    }
  } else {
    _step_delay_ms = 1000; // 默认值
  }
}

// 角度转换为步数
long StepperMotor::angleToStep(float angle) {
  // 角度转步数：angle / 360 * 步数每转
  // 对于28BYJ-48，使用半步模式时，每转4096步

  int steps_per_sequence;
  switch (_excitation) {
  case SINGLE_PHASE_FULL_STEP_EXCITATION:
    steps_per_sequence = 4;
    break;
  case TWO_PHASE_FULL_STEP_EXCITATION:
    steps_per_sequence = 4;
    break;
  case HALF_PHASE_EXCITATION:
  default:
    steps_per_sequence = 8;
    break;
  }

  float actual_steps_per_rev = FULL_ROTATION_STEPS * (steps_per_sequence / 8.0);
  return round(angle * actual_steps_per_rev / 360.0);
}

// 步数转换为角度
float StepperMotor::stepToAngle(long step) {
  int steps_per_sequence;
  switch (_excitation) {
  case SINGLE_PHASE_FULL_STEP_EXCITATION:
    steps_per_sequence = 4;
    break;
  case TWO_PHASE_FULL_STEP_EXCITATION:
    steps_per_sequence = 4;
    break;
  case HALF_PHASE_EXCITATION:
  default:
    steps_per_sequence = 8;
    break;
  }

  float actual_steps_per_rev = FULL_ROTATION_STEPS * (steps_per_sequence / 8.0);
  return step * 360.0 / actual_steps_per_rev;
}

// 检查是否到达目标角度
bool StepperMotor::reachedTargetAngle() {
  return (_mode == MODE_NONE) ||
         (_current_step == _target_step && _mode == MODE_RUN_WITH_POSITION);
}

// 获取当前步数
long StepperMotor::getCurrentStep() { return _current_step; }

// 设置当前步数
void StepperMotor::setCurrentStep(long current_step) {
  _current_step = current_step;
}

// 获取当前角度
float StepperMotor::getCurrentAngle() { return stepToAngle(_current_step); }

// 设置当前角度
void StepperMotor::setCurrentAngle(float current_angle) {
  _current_step = angleToStep(current_angle);
}

// 获取目标步数
long StepperMotor::getTargetStep() { return _target_step; }

// 设置目标步数
void StepperMotor::setTargetStep(long target_step) {
  _mode = MODE_RUN_WITH_POSITION;
  _target_step = target_step;
}

// 获取目标角度
float StepperMotor::getTargetAngle() { return _target_angle; }

// 设置目标角度
void StepperMotor::setTargetAngle(float target_angle) {
  _target_angle = target_angle;
  _mode = MODE_RUN_WITH_POSITION;
  _target_step = angleToStep(_target_angle);
}

// 移动指定步数
void StepperMotor::move(long steps) { setTargetStep(_current_step - steps); }

// 旋转指定角度
void StepperMotor::rotate(float angle) {
  setTargetAngle(getCurrentAngle() - angle);
}

// 正向旋转
void StepperMotor::forward() { _mode = MODE_RUN_FORWARD; }

// 反向旋转
void StepperMotor::backward() { _mode = MODE_RUN_BACKWARD; }

// 停止
void StepperMotor::stop() {
  _mode = MODE_NONE;
  // 将所有引脚设置为低电平以停止电机
  setPins(0, 0, 0, 0);
}

// 更新函数（需要在loop中调用）
void StepperMotor::tick() {
  if (_mode == MODE_NONE) {
    return;
  }

  unsigned long current_time = millis();

  // 检查是否到了下一步的时间
  if (current_time - _last_step_time < _step_delay_ms) {
    return;
  }

  _last_step_time = current_time;

  // 根据模式执行动作
  if (_mode == MODE_RUN_WITH_POSITION) {
    if (_target_step > _current_step) {
      // 正向旋转
      _current_step++;
      // 更新序列步数
      _current_sequence_step = (_current_sequence_step + 1) % 8;
    } else if (_target_step < _current_step) {
      // 反向旋转
      _current_step--;
      // 更新序列步数（反向）
      _current_sequence_step = (_current_sequence_step + 7) % 8;
    } else {
      _mode = MODE_NONE;
      return;
    }
  } else if (_mode == MODE_RUN_FORWARD) {
    _current_step--;
    _current_sequence_step = (_current_sequence_step + 7) % 8;
  } else if (_mode == MODE_RUN_BACKWARD) {
    _current_step++;
    _current_sequence_step = (_current_sequence_step + 1) % 8;
  }

  // 设置引脚状态
  const int *step_values = STEP_SEQUENCE[_excitation][_current_sequence_step];
  setPins(step_values[0], step_values[1], step_values[2], step_values[3]);
}

// 阻塞式运行直到完成
void StepperMotor::runBlocking() {
  if (_mode == MODE_RUN_WITH_POSITION) {
    long steps_to_move = _target_step - _current_step;
    int direction = (steps_to_move > 0) ? 1 : -1;
    steps_to_move = abs(steps_to_move);

    for (long i = 0; i < steps_to_move; i++) {
      // 执行一步
      if (direction > 0) {
        _current_sequence_step = (_current_sequence_step + 1) % 8;
        _current_step++;
      } else {
        _current_sequence_step = (_current_sequence_step + 7) % 8;
        _current_step--;
      }

      // 设置引脚状态
      const int *step_values =
          STEP_SEQUENCE[_excitation][_current_sequence_step];
      setPins(step_values[0], step_values[1], step_values[2], step_values[3]);

      // 延迟
      delay(_step_delay_ms);
    }

    _mode = MODE_NONE;
  } else if (_mode == MODE_RUN_FORWARD || _mode == MODE_RUN_BACKWARD) {
    // 对于连续旋转模式，阻塞运行没有意义
    Serial.println("警告: 连续旋转模式不能使用阻塞运行!");
  }
}

// 运行函数
void StepperMotor::run() {
  // 已经在tick中实现了非阻塞运行
  // 这个函数主要是为了API兼容性
}
